client/allocrunnerv2/alloc_runner.go

```diff
@@ -218,6 +233,159 @@ func (ar *allocRunner) Restore() error {
 	})
 }
 
+// TaskStateUpdated is called when a task's state has been updated. This hook is
+// used to compute changes to the alloc's ClientStatus and to update the server
+// with the new state.
+func (ar *allocRunner) TaskStateUpdated(taskName string, state *structs.TaskState) error {
+	if state == nil {
+		return fmt.Errorf("nil state given to TaskStateUpdated")
+	}
+
+	// If a task is dead, we potentially want to kill other tasks in the group
+	if state.State == structs.TaskStateDead {
+		// Find all tasks that are not the one that is dead and check if the one
+		// that is dead is a leader
+		var otherTaskRunners []*taskrunner.TaskRunner
+		var otherTaskNames []string
+		leader := false
+		for name, tr := range ar.tasks {
+			if name != taskName {
+				otherTaskRunners = append(otherTaskRunners, tr)
+				otherTaskNames = append(otherTaskNames, name)
+			} else if tr.Task().Leader {
+				leader = true
+			}
+		}
+
+		// If the task failed, we should kill all the other tasks in the task group.
+		if state.Failed {
+			for _, tr := range otherTaskRunners {
+				tr.Kill(context.Background(), structs.NewTaskEvent(structs.TaskSiblingFailed).SetFailedSibling(taskName))
+			}
+			if len(otherTaskRunners) > 0 {
+				ar.logger.Debug("task failure, destroying all tasks", "failed_task", taskName, "destroying", otherTaskNames)
+			}
+		} else if leader {
+			// If the task was a leader task we should kill all the other tasks.
+			for _, tr := range otherTaskRunners {
+				tr.Kill(context.Background(), structs.NewTaskEvent(structs.TaskLeaderDead))
+			}
+			if len(otherTaskRunners) > 0 {
+				ar.logger.Debug("leader task dead, destroying all tasks", "leader_task", taskName, "destroying", otherTaskNames)
+			}
+		}
+	}
+
+	// Gather the state of the other tasks
+	states := make(map[string]*structs.TaskState, len(ar.tasks))
+	for name, tr := range ar.tasks {
+		if name == taskName {
+			states[name] = state
+		} else {
+			states[name] = tr.TaskState()
+		}
+	}
+
+	// Get the client allocation
+	calloc := ar.clientAlloc(states)
+
+	// Update the server
+	if err := ar.stateUpdater.AllocStateUpdated(calloc); err != nil {
+		ar.logger.Error("failed to update remote allocation state", "error", err)
+	}
+
+	return nil
+}
+
+// clientAlloc takes in the task states and returns an Allocation populated
+// with Client specific fields
+func (ar *allocRunner) clientAlloc(taskStates map[string]*structs.TaskState) *structs.Allocation {
+	ar.stateLock.RLock()
+	defer ar.stateLock.RUnlock()
+
+	a := &structs.Allocation{
+		ID:         ar.id,
+		TaskStates: taskStates,
+	}
+
+	s := ar.state
+	if d := s.DeploymentStatus; d != nil {
+		a.DeploymentStatus = d.Copy()
+	}
+
+	// Compute the ClientStatus
+	if s.ClientStatus != "" {
+		// The client status is being forced
```

Comment:
does this happen anymore?

Reply:
from discussion: yes, alloc runner needs to be able to set task state when it encounters failures